### PR TITLE
feat: add junit2otel step supporting traces

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
It overrides the junit step, supporting passing traces if desired.

The traces are covered by a feature flag i the form of the `JUNIT_2_OTLP` variable. If not empty, the code will evaluate the traces. In any case, at the end of the step it will execute the default junit step, passing all arguments to it.

The traces will be sent using https://github.com/mdelapenya/junit2otlp CLI.

To process the traces, the step:

1. will set the service name, service version and trace name from the environment:
  - If the REPO variable is set, but the OTEL_SERVICE_NAME is not, then REPO will be used as service name. Otherwise OTEL_SERVICE_NAME will be used, and `junit2otel` in the case none of the variables are set.
  - If the REPO variable is set, but the JUNIT_OTEL_TRACE_NAME is not, then REPO will be used as trace name. Otherwise JUNIT_OTEL_TRACE_NAME will be used, and `junit2otel` in the case none of the variables are set.
  - If the JUNIT_OTEL_SERVICE_VERSION is not set, then it will use, in this particular order: pull-request ID, tag name and branch name. Else, it will use 'unknown'.
2. will call a shell script to avoid overloading the jenkins instance
3. the shell script will capture the glob expression representing the `testResults` attribute from the default junit step
4. it will loop over those files, executing the CLI tool as a Docker image.

## Why is it important?
We want to start sending traces for test reports

## How to test the OpenTelemetry traces
I was able to run the shell script having configured Jaeger previously:

1. Prepare Jaeger and OTLP exporter:
```shell
git clone https://github.commdelapenya/junit2otlp.git && cd junit2otlp
make demo-start-jaeger
```

2. Prepare environment to connect to Jaeger from the CLI tool:
```
export OTEL_EXPORTER_OTLP_INSECURE="true"
export OTEL_EXPORTER_OTLP_METRIC_INSECURE="true"
export OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:14317"
export OTEL_EXPORTER_OTLP_SPAN_INSECURE="true"
export OTEL_EXPORTER_OTLP_HEADERS=""
```

3. Run the shell script:
```
$ SERVICE_NAME="apm-pipeline" SERVICE_VERSION="master" TRACE_NAME="junit-tests" TEST_RESULTS_GLOB="/Users/mdelapenya/sourcecode/src/github.com/mdelapenya/junit2otlp/TEST-*" ./resources/scripts/junit2otel.sh 
```

4. Check results on Jaeger: open http://localhost:16686/, and traces for the `apm-pipeline` service should be displayed, using `junit-tests` as trace name and `master` as version.

Screenshot 2022-01-26 at 18 17 02

## Related issues
- Supersedes https://github.com/elastic/apm-pipeline-library/pull/1368

### CODE REVIEW

File: resources/scripts/junit2otel.sh
Comment (v1v) [inline resources/scripts/junit2otel.sh:32]: nit, can you split in lines with `/`?

```suggestion
  cmd < "$f" | docker run \
    --rm -i \
    --network host \
    --env "OTEL_EXPORTER_OTLP_ENDPOINT=${OTEL_EXPORTER_OTLP_ENDPOINT}" \
    --env "OTEL_EXPORTER_OTLP_HEADERS=${OTEL_EXPORTER_OTLP_HEADERS}" \
    ${DOCKER_IMAGE} \
    --service-name "${SERVICE_NAME}" \
    --service-version "${SERVICE_VERSION}" \
    --trace-name "${TRACE_NAME}"
```

File: vars/junit.groovy
Comment (v1v) [inline vars/junit.groovy:57]: You might need to remove the arguments that are not supported by the `junit` step, such as: serviceName, serviceVersion, otherwise there is a signature validation IIRC

File: vars/junit.groovy
Comment (v1v) [inline vars/junit.groovy:33]: also a docker validation just in case?

File: vars/junit.groovy
Comment (mdelapenya) [reply vars/junit.groovy:57]: Ah cool, I thought they were automatically discarded

**Comment by kuisathaverat** [comment]
I don't like to overwrite the junit step, this change affects all Jenkins instances using the library, and it is a particular behavior you have to configure, so I vote for a separate step to do the thing.

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:131]: You can simplify this with:

```groovy
pipeline {
  agent { label 'linux && immutable' }
  environment {
    JUNIT_2_OTLP = "true"
    OTEL_SERVICE_NAME = "apm-pipeline-library"
    JUNIT_OTEL_SERVICE_VERSION = "main"
    JUNIT_OTEL_TRACE_NAME = "junit-tests"
```

or even at the stage level, then we don't need to change the junit call with a closure:

```groovy
stage('Test') {
  environment {
    JUNIT_2_OTLP = "true"
    OTEL_SERVICE_NAME = "apm-pipeline-library"
    JUNIT_OTEL_SERVICE_VERSION = "main"
    JUNIT_OTEL_TRACE_NAME = "junit-tests"
  }
  steps {
```

File: vars/README.md
Comment (v1v) [inline vars/README.md:1907]:
```suggestion
                        // JUnit with OpenTelemetry traces
```

File: vars/junit.groovy
Comment (v1v) [inline vars/junit.groovy:28]:
```suggestion
                        // JUnit with OpenTelemetry traces
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [reply .ci/Jenkinsfile:131]: I wanted to make it explicit, although I'm fine with enabling them at the pipeline level

File: vars/junit.txt
Comment (mdelapenya) [inline vars/junit.txt:12]:
```suggestion
                        // JUnit with OpenTelemetry traces
```

File: vars/junit.groovy
Comment (v1v) [inline vars/junit.groovy:71]: NIT: elvis can reduce the code, no strong opinion :)

```suggestion
    withEnv([
      "SERVICE_NAME=${env.OTEL_SERVICE_NAME?.trim() ? env.OTEL_SERVICE_NAME?.trim() : 'junit2otlp'}",
      "SERVICE_VERSION=${env.JUNIT_OTEL_SERVICE_VERSION?.trim() ? env.JUNIT_OTEL_SERVICE_VERSION?.trim() : '0.0.0'}",
      "TEST_RESULTS_GLOB=${testResults}",
      "TRACE_NAME=${env.JUNIT_OTEL_TRACE_NAME?.trim() ? env.JUNIT_OTEL_TRACE_NAME?.trim() : 'junit2otlp'}"
    ]){
      withOtelEnv() {
        log(level: 'INFO', text: "Sending traces for '${env.SERVICE_NAME}-${env.SERVICE_VERSION}-${env.TRACE_NAME}'")
```

File: vars/junit.txt
Comment (v1v) [inline vars/junit.txt]:
```suggestion
                        // JUnit with OpenTelemetry traces
```

**Review by v1v — APPROVED** [review]
Some comments but no blocking this PR, great work! 💯

File: vars/README.md
Comment (mdelapenya) [inline vars/README.md:1926]:
```suggestion
```

File: vars/junit.txt
Comment (mdelapenya) [inline vars/junit.txt:31]:
```suggestion
```

File: .ci/Jenkinsfile
Comment (kuisathaverat) [inline .ci/Jenkinsfile:127]: to report to Otel, Do you have to pass those env vars? if so, it means that you have to edit the Jenkinsfile on those projects you want to report Otel data from a JUnit file, in that case, I do not see the advantage of overwriting the unit step.

Comment (mdelapenya) [reply .ci/Jenkinsfile:127]: We could infer those values using a company/team wide convention:

- OTEL_SERVICE_NAME == env.REPO
- JUNIT_OTEL_SERVICE_VERSION (if isPR then "use PR$ID", else if isBranch then "use branch name" else if isTag "use tag name") else "use undefined"?
- JUNIT_OTEL_TRACE_NAME == "env.REPO-junit-tests"

Then we would not need to set the env in the outer pipeline

File: vars/junit.groovy
Comment (mdelapenya) [reply vars/junit.groovy]: I'm reverting this, as it caused the tests to fail: `log(level: 'INFO', text: "Sending traces for '${env.SERVICE_NAME}-${env.SERVICE_VERSION}-${env.TRACE_NAME}'")` was printing

>Sending traces for 'null-null-null'"

Comment (v1v) [reply vars/junit.groovy:71]: Probably there is a bug in the withEnv mock test class https://github.com/elastic/apm-pipeline-library/blob/abff9fd58c7cc28b38ac77c9fe68591b5f3be85e/src/test/groovy/co/elastic/TestUtils.groovy#L22-L33

does it work if you don't use `env.` but the variable name itself?

File: .ci/Jenkinsfile
Comment (mdelapenya) [reply .ci/Jenkinsfile:127]: @kuisathaverat @v1v any suggestion to unblock this?

Comment (v1v) [reply .ci/Jenkinsfile:127]: What do you think if we split this implementation in two stages:
1) To support the OTEL junit tracing with a new shared library step. Then we test in a couple of place and see how it goes.
2) To discuss if the `junit` can be the one to use broadly this junit implementation and then apply the changes.

T

Comment (kuisathaverat) [reply .ci/Jenkinsfile:127]: I do not like it, but I am not blocking it

Comment (mdelapenya) [reply .ci/Jenkinsfile:127]: And about the predefined name convention?

I'd see it with good eyes to have another step: `junitWithOtel` and test it elsewhere

Comment (mdelapenya) [reply .ci/Jenkinsfile:127]: @kuisathaverat @v1v I've updated the code renaming the step to `junit2Otel` so that it does not override the default junit step

Comment (mdelapenya) [reply .ci/Jenkinsfile:127]: If you agree, we can merge this one

Comment (v1v) [reply .ci/Jenkinsfile:127]: Changes needed in the `.ci/Jenkinsfile` to use `junit2Otel` ?

File: .ci/Jenkinsfile
Comment (mdelapenya) [reply .ci/Jenkinsfile:131]: I added this support, updating the PR description

File: .ci/Jenkinsfile
Comment (mdelapenya) [reply .ci/Jenkinsfile:127]: Once released we can include it in the current Jenkinsfile. With latest changes, becuase the step will evaluate the env.REPO for the service/trace name, and isPR/isTag/isBranch to get the current version, the change will be simply changing the `junit` call for `junit2Otel` (⚠️ camelcase)

File: vars/junit2Otel.txt
Comment (v1v) [inline vars/junit2Otel.txt:38]: obsoleted?

**Review by kuisathaverat — APPROVED** [review]

**Comment by v1v** [comment]
@mdelapenya , there are only a couple of minor comments:
- references in the docs about the junit
- the PR title points to junit

Other than that 💯

File: vars/junit2Otel.txt
Comment (mdelapenya) [reply vars/junit2Otel.txt:38]: I think it's still needed, as we are passing the arguments to the internal junit call

File: vars/junit2Otel.groovy
Comment (v1v) [inline vars/junit2Otel.groovy:54]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:218]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:192]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:179]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:166]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:153]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:139]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:125]:
```suggestion
```

File: src/test/groovy/Junit2OtelStepTests.groovy
Comment (v1v) [inline src/test/groovy/Junit2OtelStepTests.groovy:111]:
```suggestion
```

File: vars/junit2Otel.txt
Comment (v1v) [reply vars/junit2Otel.txt:38]: I didn't see https://github.com/elastic/apm-pipeline-library/pull/1514/files#diff-c801f25bc4862b6758d33c9589b64c8e0d55b8bdf37aad7dec2e63a66a3c73caR81  :/

File: vars/junit2otel.groovy
Comment (v1v) [inline vars/junit2otel.groovy:66]:
```suggestion
      error('junit2otel: docker is not installed but required.')
```

**Review by v1v — APPROVED** [review]

File: vars/junit2otel.groovy
Comment (mdelapenya) [reply vars/junit2otel.groovy]: good catch!

**Pull request merged**
